File: 238. Product of Array Except Self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        left = [0]*len(nums)
        prod = 1
        for i in range(len(nums)):
            left[i] = prod
            prod *= nums[i]
            
        prod = 1
        for i in range(len(nums)-1, -1, -1):
            left[i] = left[i]*prod
            prod *= nums[i]
            
        return left

File: 238. Product of Array Except Self/solution.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] prod = new int[nums.length];
        
        // left to right
        prod[0] = 1;
        for (int i = 1; i < nums.length; i++) {
            prod[i] = prod[i-1] * nums[i-1];
        }
        
        // right to left
        int runningProd = nums[nums.length-1];
        for (int i = nums.length - 2; i >= 0; i--) {
            prod[i] = prod[i] * runningProd;
            runningProd *= nums[i];
        }
        
        return prod;
    }
}
